ceph-disk,osd: add support for crush device classes

dachary opened the pull request:
http://tracker.ceph.com/issues/19513

dachary commented:
@liewegas could you please take a quick look at ecf934ef250e17197cda16c4fb7b05984b498c50 ? It's passing tests and I'll split it into two commits for clarity. Unless you see something obviously wrong with it at first glance ;-)

liewegas commented:
lgtm!

liewegas commented:
The fallback to 'osd create' for both commands is sort of weird at first blush but makes sense since the crush location update (and/or device class) might be disabled.

dachary commented:
@liewegas thanks for the quick review :-) I split the commit in three for clarity and rebased after @yuriw merged the pull request it depended on. It should be ready for your final review.

liewegas commented on src/ceph-disk/ceph_disk/main.py:
```diff
@@ -1922,6 +1922,10 @@ def parser():
             help='unique OSD uuid to assign this disk to',
         )
         parser.add_argument(
+            '--crush-class',
```
should this be --crush-device-class maybe?

dachary replied:
We already have **ceph osd crush class {ls,rm,create}** and it looked vaguely consistent to have **--crush-class**. But if it feels weird I can change it.

dachary replied:
maybe we should change **osd crush set-device-class** to **osd crush set-crush-class** so they all look the same ?

dachary replied:
In the context CrushWrapper we use **device** but in the mon we use **osd**. In the context of **ceph osd crush** repeating **device** or even **crush** is redundant and we can just have **set-class** instead. I used **set-device-class** initially because the CrushWrapper function has the same name but in retrospect I think it was not a good idea.

liewegas replied:
Ok I buy that argument.  Let's go with it!

liewegas commented on src/osd/OSD.cc:
```diff
@@ -3021,6 +3021,43 @@ int OSD::shutdown()
   return r;
 }
 
+int OSD::mon_cmd_maybe_osd_create(string &cmd) {
```
nit: newline before {

dachary replied:
oops, fixed and repushed

dachary commented:
@liewegas I added  	b17989e to s/set-device-class/set-class/  and have **crush-class** when the OSD is implied or in context and **class** when both the OSD and crush are implied. It feels consistent to me right now but ...

liewegas commented:
I think it's consistent, but I worry that we have both "type" and "class" and the user can be easily confused.  If we refer to them as "bucket type" and "device class" throughout then they will be less likely to be mixed up...

dachary commented:
that's sensible. Shall we change **ceph osd crush class {rm,ls,create}** to **ceph osd crush device-class {rm,ls,create}** then ?

liewegas commented:
i'm happy with either option; you decide!  :)

dachary commented:
<pre>
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:85return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:11return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:17return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:26return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:39return 1
</pre>

jenkins test this please

dachary commented:
@liewegas I prefer **class** and **crush_class** over **device_class** :-)

liewegas commented:
:+1:

dachary commented:
repushed after renaming **crush_class** into **crush_device_class**

dachary commented:
@liewegas I think I addressed your remarks. Did I forget something ?

yuriw commented:
http://pulpito.ceph.com/yuriw-2017-04-17_21:11:42-rados-wip-sage-testing2_2017_4_18-distro-basic-smithi

http://pulpito.ceph.com/yuriw-2017-04-18_14:45:29-rados-wip-sage-testing2_2017_4_18---basic-smithi/

yuriw commented:
@liewegas pls merge
